### ISSUE TITLE
ReSharper running all tests using Machine Specifications

### ISSUE THREAD

**Issue by Cesar-Ofi** [machine/machine.specifications.runner.resharper#47]
I'm trying to debug a specific MSPEC test, but it seems ReSharper is running all the tests. My session only contains the one test, but breakpoints get triggered and I can clearly see in the stacktrace they're from another test.

After running, the session still displays just the single test I'm trying to run. I also noticed this when not debugging, for example when running a specific test which would get delayed because of other longer running tests.

Using R# Ultimate 2016.2.2 and Machine.Specifications.Runner.Resharper 1.6.0-beta5. I tried different variations, starting a fresh session by adding the test, using Run Test, using Debug Test and still the same behavior - while the session window only shows the one test after running, it's clearly running more than just the one I want.

Am I missing something or is this a bug?

**Comment by Cesar-Ofi** [comment]
New version 2016.3 installed and still same issue. Is this a R# error or is this runner?

**Comment by kevinkemp** [comment]
I have the same issue.  This is an inferior product (slower, no support for context-sensitive testing etc), but is at least workable until this is fixed: https://github.com/machine/machine.vstestadapter

**Comment by ivanz** [comment]
Hi both. Thanks for reporting this. When you say `test` - do you refer to a single `It`? There is a known limitation in MSpec where it doesn't expose a way to run a single `It`, so all tests within the class have to be executed. There is work in progress to address this within MSpec itself. I am waiting to see what will come out of all the changes in the .NET Tooling space with dotnet/VS 2017 before this is wrapped up and all runners updated to support this.

**Comment by Cesar-Ofi** [comment]
I Ivanz.

I see that it runs all tests in project, I have one project with 1500 tests in different classes, and I can see that it runs all tests (1500) when I only right click one It and select Debug.

I can see this, because all tests inherits from a base class, and if I set a breakpoint I can see in call stack that is running other contexts, not the one that I selected.

Hope this helps

**Comment by ivanz** [comment]
Can you please extract a sample and provide steps to reproduce the issue? I have been unable to reproduce.For example:

Given the below when I run `SpecA.a1` only `SpecA.a1` and `SpecA.a2` (and a2 runs due to the limitation I mentioned below), but `SpecB.b1` never runs.

```
public class SpecA
{
    It a1 = () => { };

    It a2 = () => { };
}

public class SpecB
{
    It b1 = () => { };
}
```

**Comment by kevinkemp** [comment]
public class SpecA : BaseSpec
{
    It a1 = () => { };

    It a2 = () => { };
}

public class SpecB : BaseSpec

{
    It b1 = () => { };
}


public class BaseSpec {

static inc _counter;

public BaseSpec() {

_counter++;

}

}


Hopefully no typos since I did that without an editor.  _counter
should get incremented 3 times.

**Comment by ivanz** [comment]
Thanks for that. In your example the reason why you are seeing the `BaseSpec()` being called is not because the tests are being run, but because an instance is being created of both `SpecA` and `SpecB`. This is a side-effect of how the current filtering works in MSpec.

**Comment by kevinkemp** [comment]
A consequence of that is that any setup done in a base class now slows down
a single test by a factor of x where x is how many implementations you've
made.  That will essentially make the R# test runner unusable for anyone
that uses base classes to setup data.  Also, besides that... the simple
test from above seems to take a long amount of time to run compared to
previous versions.  I don't have any concrete data to back that up, though.

**Comment by ivanz** [comment]
It's not optimal for sure that types are getting instantiated. If the setup is done in `Establish` instead of constructors - you will mitigate that problem, because only the test-to-run `Establish` should be invoked.

**Comment by Cesar-Ofi** [comment]
I have Establish in BaseSpec, my tests inherits from BaseSpec, is similar to @kevinkemp sample :(

**Comment by Cesar-Ofi** [comment]
Fixed,

My BaseSpec now have a Establish and not a simple constructor, now I can debug test by test, not all

**Comment by kevinkemp** [comment]
Yes, thanks for the tip.  I had previously used the Establish from the base class for setup and moved away from it but can't remember why.  At the very least, I now have a better reason to do it that way again.
